libraries/standard/http/src/http_client.c

```diff
@@ -1,42 +1,362 @@
+        if( strncmp( pField,
+                     HTTP_CONNECTION_FIELD, HTTP_CONNECTION_FIELD_LEN ) )
+        {
+            /* TODO: Add log. */
+            status = HTTP_INVALID_PARAMETER;
+        }
+
+        /* User must not set "Host" header through this method. */
+        if( strncmp( pField,
+                     HTTP_HOST_FIELD, HTTP_HOST_FIELD_LEN ) )
+        {
+            /* TODO: Add log. */
+            status = HTTP_INVALID_PARAMETER;
+        }
+
+        /* User must not set "User-Agent" header through this method. */
+        if( strncmp( pField,
+                     HTTP_USER_AGENT_FIELD, HTTP_USER_AGENT_FIELD_LEN ) )
+        {
+            /* TODO: Add log. */
+            status = HTTP_INVALID_PARAMETER;
+        }
+    }
+
+    if( status == HTTP_SUCCESS )
+    {
+        status = _addHeader( pRequestHeaders,
+                             pField, fieldLen, pValue, valueLen );
+    }
+
+    return status;
 }
 
-HTTPStatus_t HTTPClient_Send( const HTTPTransportInterface_t* pTransport,
-                                  const HTTPRequestHeaders_t* pRequestHeaders,
-                                  const uint8_t* pRequestBodyBuf, // For a PUT or POST request.
-                                  size_t reqBodyBufLen,
-                                  httpResponse_t* pResponse )
+HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                        int32_t rangeStart,
+                                        int32_t rangeEnd )
+{
+    /* Create buffer to fit max possible length for "bytes=<start>-<end>".
+     * This is the value of the Range header. */
+    char rangeValueStr[ HTTP_RANGE_BYTES_VALUE_MAX_LEN ] = { 0 };
+    char * pRangeValueCur = &rangeValueStr;
+    /* Excluding all the remaining null bytes. */
+    size_t rangeValueStrActualLength = 0;
+    int32_t bytesWritten = 0;
+
+    /* Write "bytes=<start>:<end>" for Range header value. */
+    memcpy( rangeValueStr,
+            HTTP_RANGE_BYTES_PREFIX_VALUE, HTTP_RANGE_BYTES_PREFIX_VALUE_LEN );
+    pRangeValueCur += HTTP_RANGE_BYTES_PREFIX_VALUE_LEN;
+    rangeValueStrActualLength += HTTP_RANGE_BYTES_PREFIX_VALUE_LEN;
+    memcpy( rangeValueStr, EQUAL_CHARACTER, EQUAL_CHARACTER_LEN );
+    pRangeValueCur += EQUAL_CHARACTER_LEN;
+    rangeValueStrActualLength += EQUAL_CHARACTER_LEN;
+    memcpy( rangeValueStr, itoa( rangeStart ), itoaLength( rangeStart ) );
+    pRangeValueCur += itoaLength( rangeStart );
+    rangeValueStrActualLength += itoaLength( rangeStart );
+    memcpy( rangeValueStr, DASH_CHARACTER, DASH_CHARACTER_LEN );
+    pRangeValueCur += DASH_CHARACTER_LEN;
+    rangeValueStrActualLength += DASH_CHARACTER_LEN;
+    bytesWritten = snprintf( rangeValueStr,
+                             HTTP_RANGE_BYTES_VALUE_MAX_LEN - rangeValueStrActualLength,
+                             "%d", rangeStart );
+    pRangeValueCur += bytesWritten;
+    rangeValueStrActualLength += bytesWritten;
```

Comment:
I believe lines 317 to 344 can be reduced to: 
```
snprintf( rangeValueStr,
          sizeof("rangeValueStr"),
          "%s=%d-",
          HTTP_RANGE_BYTES_PREFIX_VALUE,
          rangeStart)    
```
Also we may not write the "-" character if the startRange is negative and we write rangeEnd depending on rules here: https://w.amazon.com/bin/view/AWS/Mobile/IoT/Treadstone/HTTPDesign/Redesign/LowLevel/#HReceivingaresponse

So my example will be in some if/else block